### PR TITLE
Move and polish ProblemReporter interface

### PR DESCRIPTION
**Pull request by donat**
- ProblemReporter now has create() report() and throwing() methods
- ProblemReporter.throwing() methods take exception as parameter

See individual commits for separate changes.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by donat** [comment]
@bot-gradle test ACTR

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemReporter.java [hunk 1]
```diff
@@ -65,42 +65,46 @@ private DefaultProblemBuilder createProblemBuilder() {
...
-            return (RuntimeException) exception;
-        } else {
-            throw new RuntimeException(exception);
-        }
+        throw runtimeException(exception);
     }
 
-    private RuntimeException throwError(Throwable exception, Problem problem) {
-        report(problem);
+    private static RuntimeException runtimeException(Throwable exception) {
```
Comment (hegyibalint) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemReporter.java:91]: 💄We might want to denote to ourselves that this wraps something? E.g. `wrapsIntoRuntimeException`

Comment (hegyibalint) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemReporter.java:91]: Or just `wrapInto`, the return type and parameter should make it also quite trivial.

File: platforms/jvm/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/daemon/DaemonGroovyCompiler.java [hunk 2]
```diff
@@ -117,12 +117,11 @@ protected DaemonForkOptions toDaemonForkOptions(GroovyJavaJointCompileSpec spec)
             File toolsJar = jvm.getToolsJar();
             if (toolsJar == null) {
                 String contextualMessage = String.format("The 'tools.jar' cannot be found in the JDK '%s'.", jvm.getJavaHome());
-                throw problemReporter.throwing(problemSpec -> problemSpec
+                throw problemReporter.throwing(new IllegalStateException(contextualMessage), problemSpec -> problemSpec
```
Comment (hegyibalint) [inline platforms/jvm/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/daemon/DaemonGroovyCompiler.java:120]: 💭Looking at this pattern I'm pretty certain we did the right decision. This looks much intentional.

**Review by hegyibalint — APPROVED** [review]
Nice work!
I have to admit that the testing is getting a bit spaghettified because of the different API versions, but I think we will optimize out those details as we go on.

**Comment by donat** [comment]
@hegyibalint 

> I have to admit that the testing is getting a bit spaghettified because of the different API versions, but I think we will optimize out those details as we go on.

I fully agree. In fact, I've spent quite some time to clean up and create a centralized test fixture to consistently handle  different API variants. At the end I found that it's simply not worth it.

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemReporter.java (hunk 1 above)
Comment (donat) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemReporter.java:91]: We don't always wrap the exception so `maybeWrapInto` seems to be the most fitting name

**Comment by donat** [comment]
@bot-gradle merge

**Pull request merged**